### PR TITLE
fix(react-combobox): fix tabIndex for clear button in Dropdown component

### LINKED ISSUE #30584: [Bug]: Cannot tab navigate back past disabled dropdown
### Library

React Northstar / v0 (@fluentui/react-northstar)

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
    CPU: (16) x64 AMD EPYC 7763 64-Core Processor
    Memory: 58.87 GB / 62.75 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Browsers:
    Chrome: 121.0.6167.139
```

### Are you reporting Accessibility issue?

None

### Reproduction

https://codesandbox.io/p/sandbox/admiring-poitras-zhs96r?file=%2Fsrc%2Fexample.tsx%3A9%2C32&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsci50qy00063j6hpdqt5iwv%2522%252C%2522sizes%2522%253A%255B85.67398119122257%252C14.326018808777434%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsci50qx00023j6hzn19pvyd%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsci50qx00033j6h3atpnb9k%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsci50qx00053j6hnomep25b%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsci50qx00023j6hzn19pvyd%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsci50qx00013j6he4r8v8gd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clsvdflez002v3j6hdu0w8o4p%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A9%252C%2522startColumn%2522%253A32%252C%2522endLineNumber%2522%253A9%252C%2522endColumn%2522%253A32%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fexample.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clsci50qx00023j6hzn19pvyd%2522%252C%2522activeTabId%2522%253A%2522clsvdflez002v3j6hdu0w8o4p%2522%257D%252C%2522clsci50qx00053j6hnomep25b%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsci50qx00043j6hnxxdkco9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsci50qx00053j6hnomep25b%2522%252C%2522activeTabId%2522%253A%2522clsci50qx00043j6hnxxdkco9%2522%257D%252C%2522clsci50qx00033j6h3atpnb9k%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsci50qx00033j6h3atpnb9k%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

## Actual Behavior

I cannot reproduce this bug in Code sandbox, but the bug is on the React React Fluent UI website: [Fluent Dropdown (disabled)](https://react.fluentui.dev/?path=/docs/components-dropdown--default#disabled)

If you tab past a disabled Dropdown element, then use `shift + tab` to move back throughout the DOM, you cannot focus on the disabled dropdown, or continue navigating backwards throughout the DOM.

## Expected Behavior

Should be able to continue focusing back past the disabled dropdown using keyboard navigation.

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by dmytrokirpa**
## Previous Behavior

There is a problem with the Dropdown component: if you tab past a disabled Dropdown element and then use shift + tab to navigate back through the DOM, you cannot focus on the disabled dropdown or continue to navigate backwards through the DOM.

The underlying issue is the clear button on a dropdown component. When no value is selected in the dropdown, the clear button remains in the DOM with `display: none` and `tabIndex: 0`, which is incompatible with tabster.

## New Behavior

A potential solution could be remove tabIndex from the disabled/hidden buttons. Consequently, tab navigation past a disabled Dropdown element and back would work as expected.

## Related Issue(s)

- Fixes #30584

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr> | `104.984 kB`<br />`34.114 kB` | `104.989 kB`<br />`34.12 kB` | `5 B` <br />`6 B` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ReactComponents.fixture.js'>react-components: entire library</abbr> | `1.097 MB`<br />`270.957 kB` | `1.097 MB`<br />`270.963 kB` | `5 B` <br />`6 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr> | `104.398 kB`<br />`34.199 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.14 kB`<br />`20.137 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `218 kB`<br />`63.151 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `44.447 kB`<br />`14.59 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `8.39 kB`<br />`2.64 kB` |
| <samp>react-tag-picker</samp> <br /> <abbr title='bundle-size/Package.fixture.js'>@fluentui/react-tag-picker - package</abbr> | `182.176 kB`<br />`54.779 kB` |
| <samp>react-timepicker-compat</samp> <br /> <abbr title='bundle-size/TimePicker.fixture.js'>TimePicker</abbr> | `107.387 kB`<br />`35.758 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/cb6df91f534b11b772721bdc16d6d4a036394be3'>cb6df91f534b11b772721bdc16d6d4a036394be3</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">37</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">45</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">82</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">98</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Avatar-mount_base.html">733</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Avatar-mount.html">686</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Button-mount_base.html">330</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Button-mount.html">340</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Field-mount_base.html">1212</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Field-mount.html">1208</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProvider-mount_base.html">781</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProvider-mount.html">754</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-mount_base.html">87</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-mount.html">103</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">37</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">45</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">82</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">98</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/MakeStyles-mount_base.html">900</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/MakeStyles-mount.html">947</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Persona-mount_base.html">1755</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/Persona-mount.html">1805</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/SpinButton-mount_base.html">1395</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/SpinButton-mount.html">1401</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/SwatchPicker-mount_base.html">1649</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32278/perf-test-react-components/SwatchPicker-mount.html">1670</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-combobox-c2e4d514-d044-4d55-bbd4-e2b3f57c2bd3.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-combobox-c2e4d514-d044-4d55-bbd4-e2b3f57c2bd3.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=32278&commitId=ccfd3f8204d0f3cfddaae7f3f35c128ba031a3e0&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.Basic High Contrast.chromium.png|2| Changed|
|Avatar Converged.badgeMask.chromium.png|2| Changed|

</details>


  <details><summary><b>Drawer</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.overlay drawer full.chromium.png|3315| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-combobox/library/src/components/Dropdown/useDropdown.tsx [hunk 2]
```diff
@@ -104,7 +104,8 @@ export const useDropdown_unstable = (props: DropdownProps, ref: React.Ref<HTMLBu
         'aria-label': 'Clear selection',
         children: <DismissIcon />,
         // Safari doesn't allow to focus an element with this
-        tabIndex: 0,
+        // when the element is not visible (display: none) we need to set it to -1 to avoid tabster issues
+        tabIndex: showClearButton ? 0 : -1,
```
Comment (layershifter) [inline packages/react-components/react-combobox/library/src/components/Dropdown/useDropdown.tsx:108]: Can we add Cypress test for this scenario?

File: packages/react-components/react-combobox/library/src/components/Dropdown/Dropdown.cy.tsx [hunk 3]
```diff
@@ -57,3 +57,35 @@ describe('Dropdown controlling open/close state', () => {
...
+    clearButton.should('have.attr', 'tabIndex', '-1');
+  });
+
+  it('clear button should be visible and focusable when the "clearable" props is set', () => {
+    mount(<DropdownComponent clearable selectedOptions={['Cat']} />);
+
+    const clearButton = cy.get(clearButtonSelector);
+
+    clearButton.should('be.visible');
+    clearButton.should('have.attr', 'tabIndex', '0');
```
Comment (layershifter) [inline packages/react-components/react-combobox/library/src/components/Dropdown/Dropdown.cy.tsx:89]: As the original issue is about tabbing can we simulate <kbd>Tab</kbd> navigation instead? See `packages/react-examples/src/react/FocusTrapZone/e2e/FocusTrapZone.e2e.tsx`

Comment (dmytrokirpa) [reply packages/react-components/react-combobox/library/src/components/Dropdown/Dropdown.cy.tsx]: done

**Comment by dmytrokirpa** [comment]
@layershifter, I have opened an issue for [tabster](https://github.com/microsoft/tabster/issues/399) to determine whether the fix should be implemented within tabster or if we should proceed with the `tabIndex=-1` option.

**Comment by miroslavstastny** [comment]
In the repro in the linked tabster issue, if you update the tabster instance to disable tab control:
```jsx
createTabster(window, { controlTab: false });
```
The tab navigation works as expected.
In Fluent, tabster is created with `controlTab: false`, so there seems to be a different root cause.
@ling1726 Can you please help with this?

**Comment by dmytrokirpa** [comment]
> In the repro in the linked tabster issue, if you update the tabster instance to disable tab control:
> 
> ```js
> createTabster(window, { controlTab: false });
> ```
> 
> The tab navigation works as expected. In Fluent, tabster is created with `controlTab: false`, so there seems to be a different root cause. @ling1726 Can you please help with this?

Neither using `createTabster(window, { controlTab: false });` nor omitting it works for me. I'm unable to navigate past a disabled element with `zindex="0"` using the tab or shift+tab. So there might be an issue with tabster.

https://github.com/user-attachments/assets/8fc0bb3c-06d5-4f91-b28d-0827a503ac5a

**Comment by dmytrokirpa** [comment]
@ling1726, could you please review this PR once more? Alternatively, should we close it and await a fix from Tabster's end?

**Comment by dmytrokirpa** [comment]
@layershifter @ling1726, could you kindly revisit this PR? It has been open for a while now, and if you believe it's not suitable for merging, perhaps we should consider closing it.

**Pull request merged**